src/reflect/scala/reflect/internal/tpe/FindMembers.scala

```diff
@@ -13,8 +13,15 @@ trait FindMembers {
   this: SymbolTable =>
 
   /** Implementation of `Type#{findMember, findMembers}` */
-  private[internal] abstract class FindMemberBase[T](tpe: Type, name: Name, excludedFlags: Long, requiredFlags: Long) {
-    protected val initBaseClasses: List[Symbol] = tpe.baseClasses
+  private[internal] abstract class FindMemberBase[T](tpe: Type, name: Name, excludedFlags: Long, requiredFlags: Long, searchCompanions: Boolean) {
+    protected val initBaseClasses: List[Symbol] = {
+      val bcs = tpe.baseClasses
+      // For Java compilation units, a `T` or `A.T` can refer to a static inner class `T`. Static inner
+      // classes are placed in the companion object. Under `searchCompanions`, we always check the companion
+      // before proceeding to the next base class.
+      if (searchCompanions) bcs.flatMap(bc => List(bc, bc.companionModule.moduleClass))
```

Comment:
Could you add some more details for the spec from your commit message here? Maybe some variation of:

> In Java, static members are inherited, so `A.T` can refer to a static member defined in a parent of `A`. In Scala we model static members as members of the class's companion object. In addition to the inheritance class hierarchy, we must be careful to follow the implied parallel hierarchy of the companion objects.